hw/drivers/flash/enc_flash/ef_da1469x/src/hw_enc_flash.c

```diff
@@ -0,0 +1,162 @@
+    QSPIC->QSPIC_CTR_EADDR_REG = h_dev->hf_size + h_dev->hf_base_addr;
+
+    /* securely DMA hardware key from secret storage to QSPI decrypt engine */
+    dma_regs->DMA_REQ_MUX_REG |= 0xf000;
+    dma_regs->DMA7_LEN_REG = 8;
+    dma_regs->DMA7_A_START_REG = MCU_OTPM_BASE + OTP_SEGMENT_USER_DATA_KEYS +
+                                 (32 * (slot));
+    dma_regs->DMA7_B_START_REG = (uint32_t)AES_HASH->CRYPTO_KEYS_START;
+    dma_regs->DMA7_CTRL_REG = DMA_DMA7_CTRL_REG_AINC_Msk |
+                              DMA_DMA7_CTRL_REG_BINC_Msk |
+                              (MCU_DMA_BUS_WIDTH_4B << DMA_DMA7_CTRL_REG_BW_Pos) |
+                              DMA_DMA7_CTRL_REG_DMA_ON_Msk;
+    while (dma_regs->DMA7_IDX_REG != 8);
+
+    /* set OTP to standby and turn off clock */
+    da1469x_otp_set_mode(OTPC_MODE_STBY);
+    da1469x_clock_amba_disable(CRG_TOP_CLK_AMBA_REG_OTP_ENABLE_Msk);
+}
+
+static uint32_t
+get_key_size_mask(int keysize)
+{
+    uint32_t val;
+
+    switch(keysize) {
+        case 256:
+            val = 3;
+            break;
+        case 192:
+            val = 1;
+            break;
+        case 128:
+            val = 0;
+            break;
+        default:
+            val = 0;
+   }
+   return (val << AES_HASH_CRYPTO_CTRL_REG_CRYPTO_AES_KEY_SZ_Pos);
+}
+
+/**
+ * AES encrypt
+ *
+ * Encrypt the plain text with the given key
+ */
+
+void
+do_encrypt(const struct hal_flash *h_dev, uint32_t *ctr, const uint8_t *src, uint8_t *tgt, int off, int cnt)
+{
+    /* Select AES CTR, set CRYPTO_ALG_MD bits to 10 */
+    uint32_t algo_sel = (2 << AES_HASH_CRYPTO_CTRL_REG_CRYPTO_ALG_MD_Pos);
+    uint32_t ks_mask = get_key_size_mask(MYNEWT_VAL(USER_AES_KEYSIZE));
+
+    /* XXX: for now assume we are only user of crypto block */
+    da1469x_clock_amba_enable(CRG_TOP_CLK_AMBA_REG_AES_CLK_ENABLE_Msk);
+
+    /*
+     * AES_HASH_CRYPTO_CTRL_REG_CRYPTO_AES_KEXP_Msk: Key Expansion needs to be
+     * set for AES 128 CTR, but not AES 256 CTR mode
+     * Set CRYPTO_CTRL_REG to:
+     *  Enable CRYPTO_OUT_MD: Write back to memory only the final block of
+     * resulting data
+     *  Set Key Size to 256 bits
+     *  Set algo mode to CTR.
+     */
+    AES_HASH->CRYPTO_CTRL_REG = AES_HASH_CRYPTO_CTRL_REG_CRYPTO_OUT_MD_Msk |
+                                ks_mask |
+//                                AES_HASH_CRYPTO_CTRL_REG_CRYPTO_AES_KEY_SZ_Msk |
+                                algo_sel;
+
+    AES_HASH->CRYPTO_LEN_REG = ENC_FLASH_BLK;
+    AES_HASH->CRYPTO_FETCH_ADDR_REG =
+        (uint32_t)src;
+    AES_HASH->CRYPTO_DEST_ADDR_REG =
+        (uint32_t)tgt;
+
+    /* Set nonce and ctr */
+    AES_HASH->CRYPTO_MREG0_REG = ctr[0];
+    AES_HASH->CRYPTO_MREG1_REG = ctr[1];
+    AES_HASH->CRYPTO_MREG2_REG = ctr[2];
+    AES_HASH->CRYPTO_MREG3_REG = ctr[3];
+
+    assert((AES_HASH->CRYPTO_STATUS_REG & 0x01) == 1);
+
+    /* securely transfer the key from OTP area */
+    do_dma_key_tx(h_dev, MYNEWT_VAL(USER_AES_SLOT));
```

Comment:
How many slots are there? Maybe it would be interesting to also add some `#if` to check that the value is valid?

Reply:
The hw provides 8 slots. The OTP area can be locked for reading and it can only be read via secure DMA channel.